### PR TITLE
feat: Add pagination in IconController[DHIS2-16046]

### PR DESCRIPTION
**Pull request by zubaira**

### CODE REVIEW

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java [hunk 1]
```diff
@@ -60,88 +64,134 @@ void shouldCreateCustomIconWhenFileResourceExist() throws IOException {
...
+    String fileResourceId3 = createFileResource();
+    createCustomIcon(fileResourceId3, keyword, iconKey3);
+
+    JsonObject response =
+        GET("/icons?type=CUSTOM&page=1&pageSize=" + pageSize).content(HttpStatus.OK);
+
+    JsonList<JsonIcon> icons = response.getList("icons", JsonIcon.class);
+
+    assertTrue(response.has("pager"));
+    assertEquals(pageSize, icons.size());
```
Comment (teleivo) [inline dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java:167]: Please follow the tests in https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/tracker/export/ExportControllerPaginationTest.java

You need to
* add tests to show pagination is on by default, maybe get an empty page as well
* assert on pager fields to make sure the numbers are correct
* assert on fields which should not be present like `total` and `pageCount`

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java [hunk 2]
```diff
@@ -60,88 +64,134 @@ void shouldCreateCustomIconWhenFileResourceExist() throws IOException {
...
+
+    JsonObject response = DELETE(String.format("/icons/%s", iconKey1)).content();
+
+    assertEquals(
+        String.format("Icon %s deleted", iconKey1), response.getString("message").string());
+  }
+
+  @Test
+  void shouldGetIconsWithPager() throws IOException {
+
```
Comment (teleivo) [inline dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java:149]: Please remove these empty lines at the start of methods. Maybe there is an IDE setting 🤷🏻

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconRequestParams.java [hunk 3]
```diff
@@ -53,6 +54,9 @@ public class IconRequestParams {
   private Date createdEndDate;
   private Date lastUpdatedStartDate;
   private Date lastUpdatedEndDate;
+  private boolean paging = true;
+  private int pageSize = Pager.DEFAULT_PAGE_SIZE;
+  private int page;
```
Comment (teleivo) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconRequestParams.java:59]:
```suggestion
  private int page = 1;
```

These fields are user-facing as they represent the query params. The defaults show up in OpenAPI specs. To users the first `page` starts at 1 no matter how we implement it internally.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java [hunk 4]
```diff
@@ -130,15 +132,24 @@ public void getIconData(@PathVariable String key, HttpServletResponse response)
   }
 
   @GetMapping
-  public @ResponseBody List<ObjectNode> getAllIcons(IconRequestParams iconRequestParams)
+  public @ResponseBody PaginatedIconResponse getAllIcons(IconRequestParams iconRequestParams)
       throws BadRequestException {
 
     IconOperationParams iconOperationParams = iconRequestParamsMapper.map(iconRequestParams);
 
+    Pager pager = iconOperationParams.getPager();
```
Comment (teleivo) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java:140]: Why are you creating this var here? Always declare it as close to your usage. For this particular case I think you can even inline it as it does not add any value.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/jdbc/JdbcCustomIconStore.java [hunk 5]
```diff
@@ -184,4 +189,17 @@ private String buildIconQuery(
 
     return sql;
   }
+
+  private String getPaginatedQuery(
+      Pager pager, String sql, MapSqlParameterSource mapSqlParameterSource) {
+
+    sql = sql + " LIMIT :limit OFFSET :offset ";
+
+    int offset = (pager.getPage() - 1) * pager.getPageSize();
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/jdbc/JdbcCustomIconStore.java:198]: in case pagination by default is not working its that your default page in the request params is 0 and not actually 1.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/icon/PaginatedIconResponse.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.util.List;
+import lombok.Getter;
+import org.hisp.dhis.common.Pager;
+
+/**
+ * @author Zubair Asghar
+ */
+@Getter
+public class PaginatedIconResponse {
```
Comment (teleivo) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/icon/PaginatedIconResponse.java:40]: Can this live in the web package with the request params? its only ever used there right?

**Review by teleivo — COMMENTED** [review]
You need to add the change from https://github.com/dhis2/dhis2-core/pull/16211 for `total` and `pageCount` fields not to be returned in `pager`.

I think the overall design is ok for now. It's different from what we have done so far as in you are using the `pager` to pass along what page and pageSize the user wants as well as return what page and pageSize they get. That will not work if you have to implement totals or maybe even prev/nextPage. Do users need the prev/nextPage?

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconRequestParams.java (hunk 3 above)
Comment (zubaira) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconRequestParams.java:59]: fixed

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/icon/PaginatedIconResponse.java (hunk 6 above)
Comment (zubaira) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/icon/PaginatedIconResponse.java:40]: moved

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java (hunk 1 above)
Comment (zubaira) [reply dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java]: Added

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/icon/IconOperationParams.java [hunk 7]
```diff
@@ -49,6 +51,8 @@ public class IconOperationParams {
   private Date createdEndDate;
   private Date lastUpdatedStartDate;
   private Date lastUpdatedEndDate;
+  private boolean paging = true;
+  private Pager pager = PagerUtils.DEFAULT_PAGING;
```
Comment (teleivo) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/icon/IconOperationParams.java:55]: I would not share this default pager as its API allows for mutations. I don't think we can trust it in that aspect.

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java [hunk 8]
```diff
@@ -60,88 +65,182 @@ void shouldCreateCustomIconWhenFileResourceExist() throws IOException {
...
+    String fileResourceId3 = createFileResource();
+    createCustomIcon(fileResourceId3, keyword, iconKey3);
+
+    JsonObject iconResponse = GET("/icons?type=CUSTOM&page=1&pageSize=2").content(HttpStatus.OK);
+    JsonPage pager = iconResponse.get("pager", JsonPage.class);
+
+    JsonList<JsonIcon> icons = iconResponse.getList("icons", JsonIcon.class);
+
+    assertHasMember(iconResponse, "pager");
+    assertHasMember(pager, "total");
```
Comment (teleivo) [inline dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java:166]: You don't need to assert on members if you have an assertion like

`assertEquals(3, pager.getTotal());`

that assertion will fail meaningfully if the pager does not have that property.

Please check all your assertion and remove these duplicate ones.

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java [hunk 9]
```diff
@@ -60,88 +65,182 @@ void shouldCreateCustomIconWhenFileResourceExist() throws IOException {
...
+    createCustomIcon(fileResourceId3, keyword, iconKey3);
+
+    JsonObject iconResponse = GET("/icons?type=CUSTOM&paging=false").content(HttpStatus.OK);
+
+    JsonPage pager = iconResponse.get("pager", JsonPage.class);
+
+    JsonList<JsonIcon> icons = iconResponse.getList("icons", JsonIcon.class);
+
+    assertHasNoMember(iconResponse, "pager");
+    assertHasNoMember(pager, "total");
```
Comment (teleivo) [inline dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java:231]: If it has no pager then it has not total as the property is `pager.total`. Please remove this one.

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java [hunk 10]
```diff
@@ -60,88 +65,182 @@ void shouldCreateCustomIconWhenFileResourceExist() throws IOException {
...
+    createCustomIcon(fileResourceId1, keyword, iconKey1);
 
-    JsonObject response = DELETE(String.format("/icons/%s", iconKey)).content();
+    String fileResourceId2 = createFileResource();
+    createCustomIcon(fileResourceId2, keyword, iconKey2);
+
+    String fileResourceId3 = createFileResource();
+    createCustomIcon(fileResourceId3, keyword, iconKey3);
+
+    JsonObject iconResponse = GET("/icons?type=CUSTOM&page=1&pageSize=2").content(HttpStatus.OK);
```
Comment (teleivo) [inline dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java:160]: If you adjust it to

```suggestion
    JsonObject iconResponse = GET("/icons?type=CUSTOM&page=2&pageSize=2").content(HttpStatus.OK);
```

you can also assert the `pager.page` for a non-default case.

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/metadata/JsonPage.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.controller.metadata;
+
+import org.hisp.dhis.jsontree.JsonObject;
+
+/** Representation of {@link org.hisp.dhis.common.Pager}. */
+public interface JsonPage extends JsonObject {
```
Comment (teleivo) [inline dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/metadata/JsonPage.java:33]:
```suggestion
public interface JsonPager extends JsonObject {
```

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/metadata/JsonPage.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.controller.metadata;
```
Comment (teleivo) [inline dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/metadata/JsonPage.java:28]: You could put it in `package org.hisp.dhis.webapi.controller.json` as there is already a class for metadata related Json assertions.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java [hunk 13]
```diff
@@ -130,15 +134,29 @@ public void getIconData(@PathVariable String key, HttpServletResponse response)
...
+  public @ResponseBody PaginatedIconResponse getAllIcons(IconRequestParams iconRequestParams)
       throws BadRequestException {
 
     IconOperationParams iconOperationParams = iconRequestParamsMapper.map(iconRequestParams);
 
+    Pager pager = null;
+
+    if (iconRequestParams.isPaging()) {
+      long total = iconService.count(iconOperationParams);
+
```
Comment (teleivo) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java:146]: Newlines IMHO are for grouping things that belong together. I would use them more sparingly and remove at least this one.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/PaginatedIconResponse.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.List;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.hisp.dhis.common.Pager;
+
+/**
+ * @author Zubair Asghar
+ */
+@Getter
+@AllArgsConstructor
```
Comment (teleivo) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/PaginatedIconResponse.java:41]: I think we should make the fields final.

**Review by teleivo — CHANGES_REQUESTED** [review]

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java (hunk 13 above)
Comment (zubaira) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java:146]: i trust on our code formatter for these things.

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/metadata/JsonPage.java (hunk 12 above)
Comment (zubaira) [reply dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/metadata/JsonPage.java:28]: we also have metadata folder as well. So i would suggest having json folder inside this metadata folder and putting metadata related jsons into this folder

Comment (zubaira) [reply dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/metadata/JsonPage.java:28]: fixed it for now

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java (hunk 13 above)
Comment (teleivo) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java:146]: The code formatter does not add such newlines and will also not remove them. So you will need to adjust vertical whitespace.

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java [hunk 15]
```diff
@@ -60,88 +65,175 @@
...
+
+    String fileResourceId2 = createFileResource();
+    createCustomIcon(fileResourceId2, keyword, iconKey2);
+
+    String fileResourceId3 = createFileResource();
+    createCustomIcon(fileResourceId3, keyword, iconKey3);
+
+    JsonObject iconResponse = GET("/icons?type=CUSTOM&paging=false").content(HttpStatus.OK);
+
+    JsonPager pager = iconResponse.get("pager", JsonPager.class);
```
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/icon/IconControllerTest.java:221]: Please address this comment

**Review by teleivo — APPROVED** [review]

**Review by muilpp — APPROVED** [review]
